mkfs/rootdir.c

```diff
@@ -455,6 +459,65 @@ static int insert_reserved_file_extent(struct btrfs_trans_handle *trans,
 	return ret;
 }
 
+static ssize_t zlib_compress_extent(struct btrfs_inode_item *btrfs_inode,
+				    u32 sectorsize, const void *in_buf,
+				    size_t in_size, void *out_buf)
+{
+	int ret;
+	z_stream strm;
+
+	strm.zalloc = Z_NULL;
+	strm.zfree = Z_NULL;
+	strm.opaque = Z_NULL;
+	ret = deflateInit(&strm, g_compression_level);
+	if (ret != Z_OK) {
+		error("deflateInit failed: %s", strm.msg);
+		return -EINVAL;
+	}
+
+	strm.next_out = out_buf;
+	strm.avail_out = BTRFS_MAX_COMPRESSED;
+	strm.next_in = (void *)in_buf;
+	strm.avail_in = in_size;
+
+	/*
+	 * Try to compress the first sector - if it would be larger,
+	 * return -E2BIG.
+	 */
+	if (!(btrfs_stack_inode_flags(btrfs_inode) & BTRFS_INODE_COMPRESS)) {
```

Comment:
>Furthermore, if we got one 1MiB extent that its first sector is not compressible but overall it can still be compressed, then I'd say we lose the chance to save some space.

The downside to this approach is that if a file is incompressible, you end up giving it extents of 128KB rather than 1MB. Files that are already compressed are going to be more common than the worst-case scenario of a low-entropy file with a high-entropy first sector.